### PR TITLE
Make sure to clean up mapBlockSource if we've already seen the block

### PR DESCRIPTION
**Pull request by theuni**
Otherwise we may leave them dangling.

Credit TheBlueMatt.

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -2134,9 +2134,12 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             }
             bool fNewBlock = false;
             ProcessNewBlock(chainparams, pblock, true, &fNewBlock);
-            if (fNewBlock)
+            if (fNewBlock) {
                 pfrom->nLastBlockTime = GetTime();
-
+            } else {
+                LOCK(cs_main);
```
Comment (promag) [inline src/net_processing.cpp]: Nit, move the lock up and remove the one below?

**Review by promag — COMMENTED** [review]
Nit, reword `Remove existing blocks from mapBlockSource`?

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/11012/commits/3f8fa7f58b68a8ed596c62d1edb85a3455a5a724

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2390,8 +2397,12 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
         }
         bool fNewBlock = false;
         ProcessNewBlock(chainparams, pblock, forceProcessing, &fNewBlock);
-        if (fNewBlock)
+        if (fNewBlock) {
```
Comment (laanwj) [inline src/net_processing.cpp]: Would be nice to factor this out to a function, as we're repeating exactly the same code three times.

Comment (theuni) [inline src/net_processing.cpp]: Would you like me to do that here? Either way works for me.

Comment (laanwj) [inline src/net_processing.cpp]: I don't care strongly, I'm fine with having this as quick fix only. But it feels a tad brittle.

Comment (theuni) [inline src/net_processing.cpp]: I believe that @TheBlueMatt had plans to clean this up further (or remove the need) in the future, so if it's all the same, I'd prefer to leave it as-is for 0.15.

**Comment by jnewbery** [comment]
utACK 3f8fa7f58b68a8ed596c62d1edb85a3455a5a724

**Comment by promag** [comment]
utACK 3f8fa7f. Should we open an issue to clean this in the future as @theuni suggests?

**Pull request merged**